sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala

```diff
@@ -170,9 +196,27 @@ case class Add(left: Expression, right: Expression) extends BinaryArithmetic {
     if (dataType.isInstanceOf[CalendarIntervalType]) {
       input1.asInstanceOf[CalendarInterval].add(input2.asInstanceOf[CalendarInterval])
     } else {
-      numeric.plus(input1, input2)
+      val result = numeric.plus(input1, input2)
+      if (checkOverflow) {
+        val resSignum = numeric.signum(result)
+        val input1Signum = numeric.signum(input1)
+        val input2Signum = numeric.signum(input2)
+        if (resSignum != -1 && input1Signum == -1 && input2Signum == -1
+          || resSignum != 1 && input1Signum == 1 && input2Signum == 1) {
```

Comment:
why the checking is different from the codegen version?

Reply:
because since here we are using `numeric` in order for the code to be generic for alla data types, `numeric` exposes `signum` but not comparison operators, which are instead available in codegen since the datatypes are specified